### PR TITLE
Tabbed vendors 

### PR DESCRIPTION
**Pull request by SplinterGP**
## About The Pull Request
tabbed vendors, finishing code, but this should allow you to turn vendors to use tabs just by changing its product format, 

image

## Why It's Good For The Game
More organization.

## Changelog
:cl:
refactor: Vendors can now have tabs
qol: all weapons vendor now has tab and share same items, Special vendors on prep are gone.
/:cl:

### CODE REVIEW

**Comment by David-Stormwell** [comment]
But there is one they fear, in their tongue Dovahkiin, Finally-a-Good-PR-Born!

**Comment by Pariah919** [comment]
finally a good damn pr

File: code/game/objects/machinery/vending/marine_vending.dm [hunk 1]
```diff
@@ -171,44 +171,56 @@
...
+		"Explosives" = list(
+			/obj/item/explosive/mine = 5,
+			/obj/item/explosive/grenade/frag/m15 = 25,
+			/obj/item/explosive/grenade/incendiary = 25,
+			/obj/item/explosive/grenade/drainbomb = 5,
+			/obj/item/explosive/grenade/cloakbomb = 25,
+			/obj/item/storage/box/m94 = 30,
+			/obj/item/storage/box/m94/cas = 10,
+		),
+		"Especialized" = list(
```
Comment (Pariah919) [inline code/game/objects/machinery/vending/marine_vending.dm:214]:
```suggestion
		"Specialized" = list(
```
no hablamos espanol o portugues aqui

**Comment by Swagile** [comment]
good pr

but actually add a real changelog bruh

**Comment by XSlayer300** [comment]
You're not touching the HvH variants yet?

File: code/game/objects/machinery/vending/vending.dm [hunk 2]
```diff
@@ -2,13 +2,34 @@
...
 
-/datum/data/vending_product
+/datum/vending_product
 	var/product_name = "generic"
 	var/atom/product_path = null
 	var/amount = 0
 	var/price = 0
 	var/display_color = "white"
 	var/category = CAT_NORMAL
+	var/tab
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:12]: autodoc

File: code/game/objects/machinery/vending/vending.dm [hunk 3]
```diff
@@ -423,33 +407,39 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
-
-	.["premium_length"] = premium.len
+	.["tabs"] = list()
+
+	for(var/datum/vending_product/R in product_records)
+		if(R.tab && !(R.tab in .["tabs"]))
+			.["tabs"] += R.tab
+		.["displayed_records"] += list(make_record_data(R))
+
+	for(var/datum/vending_product/R in hidden_records)
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:417]: AS
?

File: code/game/objects/machinery/vending/vending.dm [hunk 4]
```diff
@@ -423,33 +407,39 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
+		.["hidden_records"] += list(make_record_data(R))
+
+	for(var/datum/vending_product/R in coin_records)
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:420]: AS
?

File: code/game/objects/machinery/vending/vending.dm [hunk 5]
```diff
@@ -423,33 +407,39 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
+		.["coin_records"] += list(make_record_data(R))
+
+/obj/machinery/vending/proc/make_record_data(datum/vending_product/R)
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:423]: Autodoc

File: code/game/objects/machinery/vending/vending.dm [hunk 6]
```diff
@@ -423,33 +407,39 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
-	for(var/datum/data/vending_product/R in hidden_records)
-		.["hidden_stock"]["[GetProductIndex(R)]"] = R.amount
-	for(var/datum/data/vending_product/R in coin_records)
-		.["coin_stock"]["[GetProductIndex(R)]"] = R.amount
-
-	.["currently_vending_name"] = currently_vending ? sanitize(currently_vending.product_name) : null
-	.["currently_vending_index"] = currently_vending_index
+	.["stock"] = list()
+
+	for(var/datum/vending_product/R in product_records + hidden_records)
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:439]: AS?

File: code/game/objects/machinery/vending/vending.dm [hunk 7]
```diff
@@ -669,7 +654,7 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
 
 //Oh no we're malfunctioning!  Dump out some product and break.
 /obj/machinery/vending/proc/malfunction()
-	for(var/datum/data/vending_product/R in src.product_records)
+	for(var/datum/vending_product/R in src.product_records)
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:657]: AS, uses src.

File: code/game/objects/machinery/vending/vending.dm [hunk 8]
```diff
@@ -691,7 +676,7 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
 	if(!target)
 		return FALSE
 
-	for(var/datum/data/vending_product/R in product_records)
+	for(var/datum/vending_product/R in product_records)
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:679]: AS

File: code/game/objects/machinery/vending/vending.dm [hunk 9]
```diff
@@ -134,47 +154,37 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
 	/obj/item/storage/pouch/firstaid
 )))
 
-/obj/machinery/vending/proc/build_inventory(list/productlist,hidden=0,req_coin=0)
+/obj/machinery/vending/proc/build_inventory(list/productlist, category = CAT_NORMAL)
+	//It's a tabbed vendor
+	if(length(productlist))
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:159]: Early return

File: code/game/objects/machinery/vending/vending.dm [hunk 10]
```diff
@@ -134,47 +154,37 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
+		var/productlistindex = productlist[1]
+		if(islist(productlist[productlistindex]))
+			for(var/tab in productlist)
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:162]: Early return

File: code/game/objects/machinery/vending/vending.dm [hunk 11]
```diff
@@ -134,47 +154,37 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
-			R.category=CAT_COIN
-			coin_records += R
+		if(isnull(amount))
+			amount = 1
+
+		var/datum/vending_product/record = new(typepath = typepath, product_amount = amount, product_price = prices[typepath], category = category)
+		if(category == CAT_HIDDEN)
+			hidden_records += record
+		else if(category == CAT_COIN)
+			coin_records += record
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:185]: Is this not a dupe of lines 168 to 173

File: code/game/objects/machinery/vending/marine_vending.dm [hunk 12]
```diff
@@ -409,9 +417,9 @@
 		stock(I, user)
 
 /obj/machinery/vending/lasgun/stock(obj/item/item_to_stock, mob/user, recharge = FALSE)
-	var/datum/data/vending_product/R //Let's try with a new datum.
+	var/datum/vending_product/R //Let's try with a new datum.
 	//More accurate comparison between absolute paths.
-	for(R in (product_records + hidden_records + coin_records))
+	for(R in (product_records + hidden_records + coin_records ))
```
[the remark this answers is not in the document]
Comment (TiviPlus) [reply code/game/objects/machinery/vending/marine_vending.dm:422]: R could also just be declared inside this for

**Comment by SplinterGP** [comment]
image
style going better, i'm a css noob

**Comment by SplinterGP** [comment]
its done

File: code/_globalvars/lists/vending.dm [hunk 13]
```diff
@@ -0,0 +1,12 @@
+/*
+* These are global lists for vending machines that all share the same inventory, it's format is simply:
+* list(typepath = list(records))
+* so all the vendors of same type have same inventory
+* it starts empty and is filled by first vendor initialized(unless its already filled).
+* the subsequent iniatilized vendors just uses the filled list.
+* there's one list for normal records, contraband and coin.
+*/
```
Comment (TiviPlus) [inline code/_globalvars/lists/vending.dm:8]:
```suggestion
/*!
 * These are global lists for vending machines that all share the same inventory, it's format is simply:
 * list(typepath = list(records))
 * so all the vendors of same type have same inventory
 * it starts empty and is filled by first vendor initialized(unless its already filled).
 * the subsequent iniatilized vendors just uses the filled list.
 * there's one list for normal records, contraband and coin.
 */
```

File: code/game/objects/machinery/vending/vending.dm [hunk 14]
```diff
@@ -23,64 +51,124 @@
...
-	var/icon_vend //Icon_state when vending!
-	var/icon_deny //Icon_state when vending!
-	var/seconds_electrified = 0 //Shock customers like an airlock.
-	var/shoot_inventory = FALSE //Fire items at customers! We're broken!
-	var/shut_up = FALSE //Stop spouting those godawful pitches!
-	var/extended_inventory = FALSE //can we access the hidden inventory?
+	///Last time we spoke our slogan
+	var/last_slogan = 0
+	//The interval between slogans.
+	var/slogan_delay = 600
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:118]:
```suggestion
	var/slogan_delay = 1 MINUTES
```

File: code/game/objects/machinery/vending/vending.dm [hunk 15]
```diff
@@ -23,64 +51,124 @@
...
-	var/slogan_delay = 600 //How long until we can pitch again?
-	var/icon_vend //Icon_state when vending!
-	var/icon_deny //Icon_state when vending!
-	var/seconds_electrified = 0 //Shock customers like an airlock.
-	var/shoot_inventory = FALSE //Fire items at customers! We're broken!
-	var/shut_up = FALSE //Stop spouting those godawful pitches!
-	var/extended_inventory = FALSE //can we access the hidden inventory?
+	///Last time we spoke our slogan
+	var/last_slogan = 0
+	//The interval between slogans.
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:117]:
```suggestion
	///The interval between slogans.
```

File: code/game/objects/machinery/vending/vending.dm [hunk 16]
```diff
@@ -23,64 +51,124 @@
...
+	var/slogan_delay = 600
+	///Icon state when successfuly vending
+	var/icon_vend
+	///Icon state when failing to vend, be it by no access or money.
+	var/icon_deny
+	///how many time we have left electrified
+	var/seconds_electrified = 0
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:124]: Ho much time?
what measure is this byond ticks or what

File: code/game/objects/machinery/vending/vending.dm [hunk 17]
```diff
@@ -134,47 +220,48 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
-/obj/machinery/vending/proc/build_inventory(list/productlist,hidden=0,req_coin=0)
-
-	for(var/typepath in productlist)
-		var/amount = productlist[typepath]
-		var/price = prices[typepath]
-		if(isnull(amount)) amount = 1
-
-		var/obj/item/temp_path = typepath
-		var/datum/data/vending_product/R = new /datum/data/vending_product()
+/obj/machinery/vending/proc/build_shared_inventory()
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:223]: Autodoc

File: code/game/objects/machinery/vending/vending.dm [hunk 18]
```diff
@@ -134,47 +220,48 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
-			R.category=CAT_COIN
-			coin_records += R
+	if(!GLOB.vending_coin_records[type])
+		build_inventory(premium, CAT_COIN)
+		GLOB.vending_coin_records[type] = coin_records
+	else
+		coin_records = GLOB.vending_coin_records[type]
+
+
+/obj/machinery/vending/proc/build_inventory(list/productlist, category = CAT_NORMAL)
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:243]: Autodoc

File: code/game/objects/machinery/vending/vending.dm [hunk 19]
```diff
@@ -134,47 +220,48 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
-			R.product_name = "packet of [SD.seed_name] [SD.seed_noun]"
-			continue
-
-		R.product_name = initial(temp_path.name)
-
+			var/amount = productlist[entry]
+			if(isnull(amount))
+				amount = 1
+			var/datum/vending_product/record = new(typepath = entry, product_amount = amount, product_price = prices[entry], category = category)
+			recordlist += record
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:264]: Use early continues

File: code/game/objects/machinery/vending/vending.dm [hunk 20]
```diff
@@ -423,33 +484,44 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
+	if(!R)
+		return
+	return list(
+		"product_name" = adminscrub(R.product_name),
+		"product_color" = R.display_color,
+		"prod_price" = R.price,
+		"prod_desc" = initial(R.product_path.desc),
+		"ref" = REF(R),
+		"tab" = R.tab,
+	)
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:515]: Redunant !R check
Cant this be a macro

File: code/game/objects/machinery/vending/vending.dm [hunk 21]
```diff
@@ -582,8 +649,9 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
 
 /obj/machinery/vending/proc/stock(obj/item/item_to_stock, mob/user, recharge = FALSE)
 	//More accurate comparison between absolute paths.
-	for(var/iter in (product_records + hidden_records + coin_records))
-		var/datum/data/vending_product/R = iter //Let's try with a new datum.
+	var/list/locallist = list()
+	locallist.Insert(product_records + hidden_records + coin_records) // this supposedly adds just the datums on a locallist so it aint nested in assocs.
+	for(var/datum/vending_product/R AS in locallist)
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:654]: Explain this, why are you using insert on a local var that just caches this

File: code/game/objects/machinery/vending/marine_vending.dm [hunk 22]
```diff
@@ -553,11 +533,11 @@
 	)
 	contraband = list()
 
-/obj/machinery/vending/MarineMed/Blood/build_inventory(productlist[])
+/obj/machinery/vending/MarineMed/Blood/build_inventory(productlist[], category)
 	. = ..()
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/marine_vending.dm:537]:
```suggestion
/obj/machinery/vending/MarineMed/Blood/build_inventory(list/productlist, category)
	. = ..()
```

**Review by TiviPlus — CHANGES_REQUESTED** [review]

File: code/game/objects/machinery/vending/vending.dm (hunk 21 above)
Comment (SplinterGP) [reply code/game/objects/machinery/vending/vending.dm:654]: i just confused myself in this probably

Comment (SplinterGP) [inline code/game/objects/machinery/vending/vending.dm:654]:
```suggestion
	for(var/datum/vending_product/R AS in product_records + hidden_records + coin_records)
```

File: code/game/objects/machinery/vending/vending.dm (hunk 16 above)
Comment (SplinterGP) [reply code/game/objects/machinery/vending/vending.dm:124]: need to check

File: code/game/objects/machinery/vending/vending.dm (hunk 20 above)
Comment (SplinterGP) [reply code/game/objects/machinery/vending/vending.dm:515]: possibly ? will look on doing

File: code/game/objects/machinery/vending/vending.dm (hunk 16 above)
Comment (SplinterGP) [reply code/game/objects/machinery/vending/vending.dm:124]: image

File: code/game/objects/machinery/vending/vending.dm [hunk 23]
```diff
@@ -23,64 +60,124 @@
...
+	var/icon_vend
+	///Icon state when failing to vend, be it by no access or money.
+	var/icon_deny
+	///how many seconds(duh) we have left electrified.
+	var/seconds_electrified = 0
+	///If we should fire items at customers! We're broken!
+	var/shoot_inventory = FALSE
+	///If true the machine won't be speaking slogans randomly. Stop spouting those godawful pitches!
+	var/shut_up = FALSE
+	///If the vending machine is hacked, makes the items on contraband list available.
```
Comment (spookydonut) [inline code/game/objects/machinery/vending/vending.dm:138]: shouldnt these be in the wires datum

Comment (TiviPlus) [reply code/game/objects/machinery/vending/vending.dm:138]: Yea but this is old code

File: code/game/objects/machinery/vending/vending.dm [hunk 24]
```diff
@@ -134,47 +229,54 @@ GLOBAL_LIST_INIT(vending_white_items, typecacheof(list(
...
-		var/datum/data/vending_product/R = new /datum/data/vending_product()
-
-		R.product_path = typepath
-		R.amount = amount
-		R.price = price
+/*!
+ * Builds shared vendors inventory
+ * the first vendor that calls this uses build_inventory and makes their records in GLOB.vending_records[type] or premium or contraband, etc.
+ * the rest of vendors of same type just set all their records to the respective global lists
+*/
```
Comment (TiviPlus) [inline code/game/objects/machinery/vending/vending.dm:236]:
```suggestion
/**
 * Builds shared vendors inventory
 * the first vendor that calls this uses build_inventory and makes their records in GLOB.vending_records[type] or premium or contraband, etc.
 * the rest of vendors of same type just set all their records to the respective global lists
 */
```
/*! is file comments

**Comment by SplinterGP** [comment]
> 
> 
> Toggling green light on a YouTool gave no contraband items like the hicap blowtorch. Might not work for other vendors with contraband. Unsure if it's by this PR but it's the closest thing.

i'll look on that later

**Comment by Losenis** [comment]
Also missing ammo boxes that were moved from RO's vendor to normal vendors

**Comment by davidchan13** [comment]
While this is an 8/10, theres a few issues though.
Green Ammo Boxes are not present (they were in the brown ammo vendors in prep/public req.)
Ammo Packets are not in the tabbed vendors (these were also in Brown Ammo Vendors.)
M-22 Bayonets can't be vended out of these (honestly you should just replace the default Bayonet with the M-22)

**Comment by TiviPlus** [comment]
Also fix many things being 1000+, it breaks with the amount left indicator

**Comment by SplinterGP** [comment]
going to fix those later

**Review by TiviPlus — APPROVED** [review]

**Comment by SplinterGP** [comment]
> 
> 
> Toggling green light on a YouTool gave no contraband items like the hicap blowtorch. Might not work for other vendors with contraband. Unsure if it's by this PR but it's the closest thing.

fix'd, was runtime

**Pull request merged**